torch/_inductor/ir.py

```diff
@@ -329,63 +342,65 @@ class IRNode:
 
     @staticmethod
     @contextlib.contextmanager
-    def current_origins(origins: Set[torch.fx.Node]):
+    def current_origins(origins: Set[Node]) -> Generator[None, None, None]:
         old = IRNode._current_origins
         IRNode._current_origins = old | origins
         try:
             yield
         finally:
             IRNode._current_origins = old
 
-    def __post_init__(self):
+    def __post_init__(self) -> None:
         self.origins = set(self._current_origins)
         self.traceback = traceback.format_stack() if config.debug_ir_traceback else None
 
     def get_read_names(self) -> Set[str]:
         raise NotImplementedError(f"NYI on {type(self)}")
 
-    def get_traceback(self):
+    def get_traceback(self) -> Optional[List[str]]:
         return self.traceback
 
-    def get_defining_op(self):
+    def get_defining_op(self) -> Optional[Operation]:
         raise NotImplementedError
 
-    def common_repr(self, shorten=True):
+    def common_repr(self, shorten: bool = True) -> Sequence[str]:
         origins = f"origins={getattr(self, 'origins', '')}"
         if shorten and len(origins) > 64:
             # this can get *very* long
             origins = f"{origins[:61]}..."
         return [origins]
 
-    def str_helper(self, lines, shorten=True, multiline=True):
-        lines = lines + self.common_repr(shorten)
+    def str_helper(
+        self, lines: Sequence[object], shorten: bool = True, multiline: bool = True
+    ) -> str:
+        lines = list(lines) + list(self.common_repr(shorten))
         lines = list(map(str, lines))
         if multiline:
             new_lines = indent(",\n".join(lines))
             return f"{type(self).__name__}(\n{new_lines}\n)"
         else:
             return f"{type(self).__name__}({lines})"
 
-    def get_dtype(self):
+    def get_dtype(self) -> torch.dtype:
         return self.dtype
 
-    def get_layout(self):
+    def get_layout(self) -> _AnyLayout:
         raise NotImplementedError(f"get_layout() is not implemented by {type(self)}!")
 
-    def get_size(self):
+    def get_size(self) -> Sequence[_IntLike]:
         raise NotImplementedError(f"get_size() is not implemented by {type(self)}!")
 
     @property
-    def shape(self):
+    def shape(self) -> Union[_IntLike, sympy.Rel, Sequence[_IntLike]]:
         return self.get_size()
```

Comment:
is sympy.Rel not subtype of `_Intlike` ? also, when will size not be sequence ?

Reply:
It is not - `issubclass(sympy.Rel, sympy.Expr)` is `False`.

For non-sequence:
`ExternKernel.realize_input` says:
```
        if isinstance(x, (Expr, sympy.logic.boolalg.Boolean, int)):
            return ShapeAsConstantBuffer(x)  # type: ignore[arg-type]
```
so if `x` is an `Expr` then it calls `ShapeAsConstantBuffer(x)` which sets shape to `x` - a non sequence.

This happens in
```
python test/inductor/test_aot_inductor.py -k test_dynamic_scalar_abi_compatible_cpu
```